File: Leetcode-Submissions/0442-find-all-duplicates-in-an-array/solution.cpp
class Solution {
public:
    vector<int> findDuplicates(vector<int>& nums) {
        vector<int> ans;

        for(int i = 0; i<nums.size(); ++i){

            int valueAtI = abs(nums[i]); // getting the absolute value at i, as it might be negative

            if(nums[valueAtI-1] < 0){

                ans.push_back(valueAtI);

            }

            nums[valueAtI-1] = nums[valueAtI-1] * -1;

        }

        return ans;
    }
};

File: Leetcode-Submissions/0442-find-all-duplicates-in-an-array/solution.java
class Solution {
    public List<Integer> findDuplicates(int[] nums) {
        int n = nums.length;
        int i = 0;
        while(i < n){
                int correctIndex = nums[i] - 1;
                if(nums[i] != nums[correctIndex]){
                    swap(nums, i, correctIndex);
                }
            else i++;
        }
       List<Integer> ans = new ArrayList<>();
        for (int index = 0; index < nums.length; index++) {
            // If the number at this index doesn't belong here, it's a duplicate
            if (nums[index] != index + 1) {
                ans.add(nums[index]);
            }
        }

        return ans;
    }
    static void swap(int[] nums, int start, int end){
        int temp = nums[start];
        nums[start] = nums[end];
        nums[end] = temp;
    }
}
